File: lib/basic/interner.ml
type t = {
    table : (string, int) Hashtbl.t
  ; mutable strings : string array
  ; next_id : int ref
}

type ident = int

let base_buf_size = 1024

let create () =
  { table = Hashtbl.create base_buf_size; strings = [||]; next_id = ref 0 }

let intern t s =
  try Hashtbl.find t.table s
  with Not_found ->
    let id = !(t.next_id) in
    Hashtbl.add t.table s id;
    t.strings <- Array.append t.strings [| s |];
    incr t.next_id;
    id

let lookup_opt t id =
  if id >= 0 && id < !(t.next_id) then Some t.strings.(id) else None

let clear t =
  Hashtbl.clear t.table;
  t.strings <- [||];
  t.next_id := 0

let size t = !(t.next_id)

File: src/basic/interner.rs
use std::collections::HashMap;

#[derive(Debug, Default)]
pub struct Interner {
    table: HashMap<String, u32>,
    strings: Vec<String>,
}

impl Interner {
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    pub fn intern(&mut self, text: &str) -> u32 {
        if let Some(&id) = self.table.get(text) {
            return id;
        }
        let id = u32::try_from(self.strings.len()).expect("interner overflow");
        let owned = text.to_owned();
        self.strings.push(owned.clone());
        let _: Option<u32> = self.table.insert(owned, id);
        id
    }

    #[must_use]
    pub fn lookup(&self, id: u32) -> Option<&str> {
        self.strings
            .get(usize::try_from(id).ok()?)
            .map(String::as_str)
    }

    pub fn clear(&mut self) {
        self.table.clear();
        self.strings.clear();
    }

    #[must_use]
    pub const fn len(&self) -> usize {
        self.strings.len()
    }

    #[must_use]
    pub const fn is_empty(&self) -> bool {
        self.strings.is_empty()
    }
}
